### ISSUE TITLE
Pinot server replacement in Cloud where we have no control over hostname

### ISSUE THREAD

**Issue by chenboat** [apache/incubator-pinot#4525]
Pinot server instance ids are formatted as Server_<host>_<port>. These ids are then used in places like table external_view and idealstate. When Pinot brokers construct the routing tables for a query, they use the above instance ids to reconstruct the host/port info to send the Pinot query via HTTP protocol. 

Right now we try to deploy Pinot in cloud here in Uber. When we replace a machine, we can not get the same hostname of the replaced one. By the current impl, the server instance id will also change due to the changed hostname. That implies the new machine can not be a replacement for the old instance even it has all the data copied to it.

We wonder how machine replacement is done in a cloud environment. One proposal we think of is to decouple the server instance id with the underlying communication protocol (e.g., HTTP used for now). In this design, the broker should NOT use the server instance id directly and instead using it to look up from Zookeeper to find out the actual server address. In fact right now these are stored in zk as HELIX_HOST and HELIX_PORT but they are apparently not used in routing table construction. 

@fx19880617 @mcvsubbu

**Comment by chenboat** [comment]
(1) Pinot broker should use the server instance id to look up the server address from Helix/Zk instead of parsing the id string like the following:

https://github.com/apache/incubator-pinot/blob/6e209e4e0ef138f8c21d597aeb20d1993fd1af2b/pinot-core/src/main/java/org/apache/pinot/core/transport/Server.java#L49-L53

(2) Pinot server should decouple the configuration of server Id with server host/port. Right now they are coupled:
https://github.com/apache/incubator-pinot/blob/3236e40c2f8d3d603c82e706db41bf2c0c2eaf68/pinot-server/src/main/java/org/apache/pinot/server/starter/helix/HelixServerStarter.java#L142-L151

**Comment by chenboat** [comment]
Currently when a Pinot server starts and joins the Helix cluster, it calls HelixManager's connect() method:
https://github.com/apache/incubator-pinot/blob/3236e40c2f8d3d603c82e706db41bf2c0c2eaf68/pinot-server/src/main/java/org/apache/pinot/server/starter/helix/HelixServerStarter.java#L160

The above method in its turn call ParticipantManager.joinCluster()

https://github.com/apache/helix/blob/7ac3a6a1762bc74869a8f7ea0a9c7feed8b3ba0a/helix-core/src/main/java/org/apache/helix/manager/zk/ParticipantManager.java#L144-L156

Notice that in the above code snippet, Helix parses the instanceName (i.e.,) into HOST and PORT field and then stores in Zk. Basically Helix couples instanceName with HELIX_HOST and HELIX_PORT.

**Comment by icefury71** [comment]
Here's how we can plan our PRs (based on Ting's recommendation)

1) PR to decouple using instance name/ID for routing requests
2) PR to change HelixServerStarter to accept a logical instance iD, host and port for starting a server.

@kishoreg @mayankshriv any thoughts on this before I proceed ?

**Comment by kishoreg** [comment]
> Here's how we can plan our PRs (based on Ting's recommendation)
> 
> 1. PR to decouple using instance name/ID for routing requests (basically don't assume instance name/ID is a valid discoverable host name. Instead use this iD to look up the real host name)
This can be done asap.
> 2. PR to change HelixServerStarter to accept a logical instance iD, host and port for starting a server.
host and port are optional, right?. If the instanceId already exists in Helix, we should change the host/port in the instanceconfig rt.
> 
> @kishoreg @mayankshriv any thoughts on this before I proceed ?

**Comment by kishoreg** [comment]
Unrelated to this PR, it will be great if we can support a mode where a node starts and grabs any available instanceId. This will be useful in cloud/kubernetes deployment where a new pinot server can provisioned and started on any host.

**Comment by icefury71** [comment]
> Unrelated to this PR, it will be great if we can support a mode where a node starts and grabs any available instanceId. This will be useful in cloud/kubernetes deployment where a new pinot server can provisioned and started on any host.

@kishoreg Yes, that can be an extension to this issue. We can have 2 ways of assigning instance ID:
a) Explicitly in HelixServerStarter (this is very useful in our environment wherein the instanceId is centrally coordinated outside of Pinot and explicitly assigned)
b) Auto assign via some discovery mechanism (either by Pinot controller or external component).

For now, are you ok with this approach ?

**Comment by kishoreg** [comment]
Looks good to me.

**Comment by icefury71** [comment]
@kishoreg thinking more about PR (1):

A performance side effect of decoupling is that every single reference to instance name will result in a Zk call (naive implementation). I'm guessing this is why the instance name was constructed with this information in the first place (which is a nice hack).

For a clean design, how about we create a helper class which caches the map of instance name to <host,port> info ? We can implement a Helix spectator to keep this map updated.

**Comment by kishoreg** [comment]
Broker already listens to the changes in instance config and caches it. This should be fairly straightforward.

**Comment by icefury71** [comment]
Created PR (1) : https://github.com/apache/incubator-pinot/pull/4778

**Comment by xiangfu0** [comment]
Closing — this has been implemented in PR #4995. Server instance ID is now decoupled from hostname/port config, enabling cloud deployments where hostnames are not controllable.
